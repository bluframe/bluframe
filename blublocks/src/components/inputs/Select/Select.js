// @flow
/* eslint-disable no-ternary */

import {
  Control,
  Input,
  Label,
  Option,
  Options,
  Triangle,
  Wrapper
} from "./styled"
import type { ComponentProps } from "."
import React from "react"

const Select = ({
  className,
  isOpen,
  label,
  name,
  onSelect,
  onToggleIsOpen,
  options,
  placeholder,
  selected
}: ComponentProps): React$Node => (
  <Wrapper className={className}>
    <Control onClick={onToggleIsOpen}>
      <Label isSelected={Boolean(selected)}>{label || placeholder}</Label>
      <Triangle isOpen={isOpen} />
    </Control>
    {isOpen ? (
      <Options>
        {options.map((option) => (
          <Option key={option.value} onClick={onSelect}>
            {option.label}
          </Option>
        ))}
      </Options>
    ) : null}
    <Input name={name} value={selected || ""} />
  </Wrapper>
)

export default Select
