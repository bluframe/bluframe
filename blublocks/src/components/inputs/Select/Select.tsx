/* eslint-disable max-lines-per-function, no-ternary */

import {
  Control,
  DisplayValue,
  Input,
  Label,
  Option,
  Options,
  Triangle,
  Wrapper
} from "./styled"
import { ForwardedRef, forwardRef } from "react"
import { ComponentProps } from "."

const Select = forwardRef(
  (
    {
      className,
      defaultValue,
      displayValue,
      isControlled,
      isLabelShrunk,
      isOpen,
      label,
      name,
      onBlur,
      onChange,
      onToggleIsOpen,
      options,
      value,
      wrapperRef
    }: ComponentProps,
    ref: ForwardedRef<HTMLInputElement>
  ) => (
    <Wrapper
      className={className}
      onBlur={onBlur}
      ref={wrapperRef}
      tabIndex={-1}
    >
      <Control onClick={onToggleIsOpen}>
        <Label $isShrunk={isLabelShrunk} id={`${name}-label`}>
          {label}
        </Label>
        <DisplayValue
          $isSelected={Boolean((!isControlled && defaultValue) || value)}
        >
          {displayValue}
        </DisplayValue>
        <Triangle $isOpen={isOpen} />
      </Control>
      {isOpen ? (
        <Options
          aria-labelledby={`${name}-label`}
          id={`${name}-listbox`}
          role="listbox"
        >
          {options.map((option, index) => (
            <Option
              aria-selected={
                (!isControlled && option.value === defaultValue) ||
                option.value === value
              }
              id={`${name}-listbox-${index}`}
              key={option.value}
              onClick={onChange(option.value)}
              role="option"
            >
              {option.label}
            </Option>
          ))}
        </Options>
      ) : null}
      <Input
        defaultValue={defaultValue}
        name={name}
        ref={ref}
        value={value === null ? undefined : value}
      />
    </Wrapper>
  )
)

Select.displayName = "Select"

export default Select
