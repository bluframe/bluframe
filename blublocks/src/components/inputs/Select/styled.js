// @flow
/* eslint-disable no-ternary */

import styled, { type StyledComponent } from "styled-components"
import type { Theme } from "@bluframe/blublocks"

export const Wrapper: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  min-width: 150px;
  position: relative;
`

export const Control: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  align-items: center;
  background: ${({ theme }) => theme.palette.text.contrast};
  border: 1px solid ${({ theme }) => theme.colors.gray.light};
  border-radius: 2px;
  column-gap: 20px;
  cursor: pointer;
  display: grid;
  grid-template-columns: auto 20px;
  padding: 5px 10px;
  height: 44px;
`

export const Label: StyledComponent<
  {| +isSelected: boolean |},
  Theme,
  HTMLDivElement
> = styled.div`
  color: ${({ isSelected, theme }) =>
    isSelected ? theme.palette.text.main : theme.palette.text.light};
`

export const Triangle: StyledComponent<
  {| +isOpen: boolean |},
  Theme,
  HTMLDivElement
> = styled.div`
  border-color: ${({ theme }) => theme.colors.gray.light};
  border-style: solid;
  border-width: 3.5px 3.5px 0 0;
  display: inline-block;
  height: 12px;
  position: relative;
  top: ${({ isOpen }) => (isOpen ? "2px" : "-2px")};
  width: 12px;
  transform: rotate(${({ isOpen }) => (isOpen ? "-45deg" : "135deg")});
`

export const Options: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  position: absolute;
  width: 100%;
`

export const Option: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  background: ${({ theme }) => theme.palette.text.contrast};
  color: ${({ theme }) => theme.palette.text.main};
  cursor: pointer;
  padding: 5px 10px;
  transition: 0.25s ease-in-out;

  &:hover {
    background: ${({ theme }) => theme.palette.primary.main};
    color: ${({ theme }) => theme.palette.primary.contrast};
    width: 100%;
  }
`

export const Input: StyledComponent<
  {||},
  Theme,
  HTMLInputElement
> = styled.input.attrs({ type: "hidden" })``
