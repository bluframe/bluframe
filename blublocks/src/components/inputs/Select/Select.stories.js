// @flow

import Select, { type Props } from "components/inputs/Select"
import React from "react"
import type { Story } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const stories = {
  component: Select,
  title: "Inputs/Select"
}

const Template = (args: Props) => <Select {...args} />

const options = [
  {
    label: "Apple",
    value: "apple"
  },
  {
    label: "Orange",
    value: "orange"
  },
  {
    label: "Banana",
    value: "banana"
  },
  {
    label: "Lemon",
    value: "lemon"
  },
  {
    label: "Pear",
    value: "pear"
  },
  {
    label: "Kiwi",
    value: "kiwi"
  },
  {
    label: "Mango",
    value: "mango"
  }
]

const args = {
  label: "Fruit",
  name: "fruit",
  onBlur: action("onBlur"),
  onChange: action("onChange"),
  options
}

export const Default: Story<Props> = Template.bind({})
Default.args = args

export const Placeholder: Story<Props> = Template.bind({})
Placeholder.args = {
  ...args,
  placeholder: "Select a fruit"
}

export const DefaultValue: Story<Props> = Template.bind({})
DefaultValue.args = {
  ...args,
  defaultValue: "orange",
  placeholder: "Select a fruit"
}

export const Value: Story<Props> = Template.bind({})
Value.args = {
  ...args,
  placeholder: "Select a fruit",
  value: "orange"
}

export default stories
