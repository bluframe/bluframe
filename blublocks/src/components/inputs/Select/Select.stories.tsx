import Select from "components/inputs/Select"
import { StoryObj } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const meta = {
  component: Select,
  title: "Inputs/Select"
}

type Story = StoryObj<typeof Select>

const options = [
  {
    label: "Apple",
    value: "apple"
  },
  {
    label: "Orange",
    value: "orange"
  },
  {
    label: "Banana",
    value: "banana"
  },
  {
    label: "Lemon",
    value: "lemon"
  },
  {
    label: "Pear",
    value: "pear"
  },
  {
    label: "Kiwi",
    value: "kiwi"
  },
  {
    label: "Mango",
    value: "mango"
  }
]

const args = {
  label: "Fruit",
  name: "fruit",
  onBlur: action("onBlur"),
  onChange: action("onChange"),
  options
}

export const Default: Story = {
  args
}

export const Placeholder: Story = {
  args: {
    ...args,
    placeholder: "Select a fruit"
  }
}

export const DefaultValue: Story = {
  args: {
    ...args,
    defaultValue: "orange",
    placeholder: "Select a fruit"
  }
}

export const Value: Story = {
  args: {
    ...args,
    placeholder: "Select a fruit",
    value: "orange"
  }
}

export default meta
