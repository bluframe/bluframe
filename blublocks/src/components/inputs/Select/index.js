// @flow
/* eslint-disable max-lines-per-function */

import type { OptionValue, SelectProps } from "@bluframe/blublocks"
import { prepareComponent, useToggle } from "@bluframe/grapple"
import { useMemo, useRef } from "react"
import Select from "./Select"

export type Props = {|
  ...SelectProps
|}

export type ComponentProps = {|
  ...SelectProps,
  +displayValue: string | React$Node,
  +isControlled: boolean,
  +isLabelShrunk: boolean,
  +isOpen: boolean,
  +onChange: (value: OptionValue | null) => () => void,
  +onToggleIsOpen: () => void,
  +wrapperRef: { current: null | HTMLDivElement }
|}

const DEFAULT_IS_OPEN = false

const usePrepareComponent = (props: Props): ComponentProps => {
  const [isOpen, toggleIsOpen] = useToggle(DEFAULT_IS_OPEN)

  // Determine if the component is controlled
  const isControlled = props.value !== undefined

  const wrapperRef = useRef(null)

  // Get displayValue and isLabelShrunk
  const [displayValue, isLabelShrunk] = useMemo(() => {
    // Get selected option
    const selectedOption = props.options.find(
      (option) =>
        (!isControlled && option.value === props.defaultValue) ||
        option.value === props.value
    )

    // If selected option exists, return its label and shrink the label
    if (selectedOption) {
      return [selectedOption.label, true]
    }

    // If no selected option and placeholder exists, return placeholder and shrink the label
    if (props.placeholder) {
      return [props.placeholder, true]
    }

    // Otherwise, return the label and shrink it when options are open
    return ["", isOpen]
  }, [
    props.defaultValue,
    props.options,
    props.value,
    props.placeholder,
    isControlled,
    isOpen
  ])

  const onBlur = (event: SyntheticEvent<*>) => {
    if (props.onBlur) {
      props.onBlur(event)
    }
  }

  const onToggleIsOpen = () => {
    // If the dropdown is currently open and is being closed by a click, trigger onBlur
    if (isOpen) {
      if (wrapperRef.current) {
        wrapperRef.current.blur()
      }
    }

    // Toggle options dropdown
    toggleIsOpen()
  }

  const onChange = (value: OptionValue | null) => () => {
    // Select via onSelect from props
    props.onChange(value)

    // If open close options dropdown
    if (isOpen) {
      onToggleIsOpen()
    }
  }

  return {
    ...props,
    displayValue,
    isControlled,
    isLabelShrunk,
    isOpen,
    onBlur,
    onChange,
    onToggleIsOpen,
    wrapperRef
  }
}

export default (prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Select
): React$ComponentType<Props>)
