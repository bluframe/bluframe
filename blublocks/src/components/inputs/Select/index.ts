/* eslint-disable max-lines-per-function */

import React, {
  FocusEventHandler,
  ForwardedRef,
  RefObject,
  useMemo,
  useRef
} from "react"
import { prepareComponent, useToggle } from "@bluframe/grapple"
import Select from "./Select"

export type OptionValue = number | string

export interface Option {
  label: React.ReactNode
  value: OptionValue
}

export interface Props {
  defaultValue?: string
  className?: string
  label: React.ReactNode
  name: string
  onBlur?: FocusEventHandler
  // eslint-disable-next-line no-unused-vars
  onChange: (value: OptionValue | null) => void
  options: Option[]
  placeholder?: string
  ref?: ForwardedRef<HTMLInputElement>
  value?: OptionValue | null
}

export interface ComponentProps extends Props {
  displayValue: React.ReactNode
  isControlled: boolean
  isLabelShrunk: boolean
  isOpen: boolean
  // eslint-disable-next-line no-unused-vars
  onChange: (value: OptionValue | null) => () => void
  onToggleIsOpen: () => void
  wrapperRef: RefObject<HTMLDivElement>
}

const DEFAULT_IS_OPEN = false

const usePrepareComponent = (props: Props): ComponentProps => {
  const [isOpen, toggleIsOpen] = useToggle(DEFAULT_IS_OPEN)

  // Determine if the component is controlled
  const isControlled = props.value !== undefined

  const wrapperRef = useRef<HTMLDivElement>(null)

  // Get displayValue and isLabelShrunk
  const [displayValue, isLabelShrunk] = useMemo(() => {
    // Get selected option
    const selectedOption = props.options.find(
      (option) =>
        (!isControlled && option.value === props.defaultValue) ||
        option.value === props.value
    )

    // If selected option exists, return its label and shrink the label
    if (selectedOption) {
      return [selectedOption.label, true]
    }

    // If no selected option and placeholder exists, return placeholder and shrink the label
    if (props.placeholder) {
      return [props.placeholder, true]
    }

    // Otherwise, return the label and shrink it when options are open
    return ["", isOpen]
  }, [
    props.defaultValue,
    props.options,
    props.value,
    props.placeholder,
    isControlled,
    isOpen
  ])

  const onBlur: FocusEventHandler = (event) => {
    if (props.onBlur) {
      props.onBlur(event)
    }
  }

  const onToggleIsOpen = () => {
    // If the dropdown is currently open and is being closed by a click, trigger onBlur
    if (isOpen) {
      if (wrapperRef.current) {
        wrapperRef.current.blur()
      }
    }

    // Toggle options dropdown
    toggleIsOpen()
  }

  const onChange = (value: OptionValue | null) => () => {
    // Select via onSelect from props
    props.onChange(value)

    // If open close options dropdown
    if (isOpen) {
      onToggleIsOpen()
    }
  }

  return {
    ...props,
    displayValue,
    isControlled,
    isLabelShrunk,
    isOpen,
    onBlur,
    onChange,
    onToggleIsOpen,
    wrapperRef
  }
}

export default prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Select
)
