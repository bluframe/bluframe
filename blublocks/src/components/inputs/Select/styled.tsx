/* eslint-disable no-ternary */

import styled from "styled-components"
import type { Theme } from "styles/theme"

export const Wrapper = styled.div`
  min-width: 150px;
  position: relative;
`

export const Control = styled.div`
  align-items: center;
  background: ${({ theme }: { theme: Theme }) => theme.palette.text.contrast};
  border: 1px solid ${({ theme }: { theme: Theme }) => theme.colors.gray.light};
  border-radius: 2px;
  column-gap: 20px;
  cursor: pointer;
  display: grid;
  grid-template-columns: auto 20px;
  padding: 5px 10px 5px 16px;
  height: 44px;
`

export const DisplayValue = styled.label`
  color: ${({ $isSelected, theme }: { $isSelected: boolean; theme: Theme }) =>
    $isSelected ? theme.palette.text.main : theme.palette.text.light};
  font-size: 16px;
  font-weight: 400;
  margin: 0;
`

export const Label = styled.label<{ $isShrunk?: boolean }>`
  color: ${({ theme }: { theme: Theme }) => theme.palette.text.light};
  font-size: 16px;
  left: 16px;
  pointer-events: none;
  position: absolute;
  top: 8px;
  transition: all 200ms;

  ${({ $isShrunk }) =>
    $isShrunk &&
    `
    font-size: 12px;
    left: 16px;
    top: -24px;
  `}
`

export const Triangle = styled.div<{ $isOpen: boolean }>`
  border-color: ${({ theme }: { theme: Theme }) => theme.colors.gray.light};
  border-style: solid;
  border-width: 3.5px 3.5px 0 0;
  display: inline-block;
  height: 12px;
  position: relative;
  top: ${({ $isOpen }) => ($isOpen ? "2px" : "-2px")};
  width: 12px;
  transform: rotate(
    ${({ $isOpen }: { $isOpen: boolean }) => ($isOpen ? "-45deg" : "135deg")}
  );
`

export const Options = styled.ul`
  list-style: none;
  margin: 0;
  max-height: 200px;
  overflow-y: auto;
  position: absolute;
  width: 100%;
  z-index: 10;
`

export const Option = styled.li`
  background: ${({ theme }: { theme: Theme }) => theme.palette.text.contrast};
  color: ${({ theme }: { theme: Theme }) => theme.palette.text.main};
  cursor: pointer;
  padding: 5px 10px;
  transition: 0.25s ease-in-out;

  &:hover {
    background: ${({ theme }: { theme: Theme }) => theme.palette.primary.main};
    color: ${({ theme }: { theme: Theme }) => theme.palette.primary.contrast};
    width: 100%;
  }
`

export const Input = styled.input.attrs({ type: "hidden" })``
