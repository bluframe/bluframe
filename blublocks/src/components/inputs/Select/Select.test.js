// @flow

import type { ComponentProps } from "."
import React from "react"
import Select from "./Select"
import { render } from "tests"

jest.mock(".", () => ({}))

describe("Select", () => {
  const props: ComponentProps = {
    displayValue: "",
    isLabelShrunk: false,
    isOpen: false,
    label: "",
    name: "fruit",
    onSelect: jest.fn(),
    onToggleIsOpen: jest.fn(),
    options: [
      {
        label: "Apple",
        value: "apple"
      },
      {
        label: "Orange",
        value: "orange"
      }
    ],
    selected: null
  }

  it("renders", () => {
    const { container } = render(<Select {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
