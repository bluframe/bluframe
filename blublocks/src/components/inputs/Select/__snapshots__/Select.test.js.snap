// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders 1`] = `
<div
  class="styled__Wrapper-sc-nm33db-0 jdNAoM"
>
  <div
    class="styled__Control-sc-nm33db-1 fjGPPF"
  >
    <div
      class="styled__Label-sc-nm33db-2 hSqRti"
    />
    <div
      class="styled__Triangle-sc-nm33db-3 fAGIZP"
    />
  </div>
  <input
    class="styled__Input-sc-nm33db-6"
    name="fruit"
    type="hidden"
    value=""
  />
</div>
`;
